### PR TITLE
Neighbor root histories

### PR DESCRIPTION
**Pull request by KeltonMad**

### CODE REVIEW

File: test/helpers.js [hunk 1]
```diff
@@ -86,6 +106,18 @@ const nonceAndId = (nonce, id) => {
   return Ethers.utils.hexZeroPad(Ethers.utils.hexlify(nonce), 8) + Ethers.utils.hexZeroPad(Ethers.utils.hexlify(id), 1).substr(2)
 }
 
+function generateDeposit(targetChainID = 0, secret = 31) {
```
Comment (nepoche) [inline test/helpers.js:109]: Do we expect deposits to use a variable amount of randomness for the secret?

Comment (KeltonMad) [reply test/helpers.js:109]: Yes. We could do this differently, but for now this works and makes sure there are not repeat commitments

File: test/integration/historicalRootWithdraw.js [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+        originDepositData,
+        resourceID,
+        { from: relayer1Address }
+    ));
+
+    /*
+    *  generate proof
+    */
+    // insert two commitments into the tree
+    await tree.insert(firstOriginDeposit.commitment);
```
[the remark this answers is not in the document]
Comment (KeltonMad) [reply test/integration/historicalRootWithdraw.js:176]: Yes

Comment (nepoche) [reply test/integration/historicalRootWithdraw.js:176]: Yeah I didn't see it further down. Maybe a test for two subsequent deposits and proof generation of the 1st with the merkle root containing the second is worthwhile?

Comment (KeltonMad) [reply test/integration/historicalRootWithdraw.js:176]: This test contains that unless I misunderstand you

File: contracts/anchors/bridged/AnchorPoseidon2.sol [hunk 3]
```diff
@@ -111,7 +113,7 @@ abstract contract AnchorPoseidon2 is MerkleTreePoseidon, ReentrancyGuard {
   */
   function withdraw(
     bytes calldata _proof,
-    bytes32 _root,
+    bytes32[2] calldata _roots,
```
Comment (drewstone) [inline contracts/anchors/bridged/AnchorPoseidon2.sol:116]: this should be a generic bytes string, and we'll parse it in the function.
```suggestion
    bytes calldata _roots,
```

File: contracts/anchors/bridged/AnchorPoseidon2.sol [hunk 4]
```diff
@@ -120,28 +122,24 @@ abstract contract AnchorPoseidon2 is MerkleTreePoseidon, ReentrancyGuard {
   ) external payable nonReentrant {
     require(_fee <= denomination, "Fee exceeds transfer value");
     require(!nullifierHashes[_nullifierHash], "The note has been already spent");
-    require(isKnownRoot(_root), "Cannot find your merkle root"); // Make sure to use a recent one
+    require(isKnownRoot(_roots[0]), "Cannot find your merkle root");
+		require(_roots.length >= edgeList.length + 1, "Incorrect root array length");
```
Comment (drewstone) [inline contracts/anchors/bridged/AnchorPoseidon2.sol:126]: spacing is off

Comment (drewstone) [reply contracts/anchors/bridged/AnchorPoseidon2.sol:126]: also we should parse the new variable as we discussed:
```
bytes32[2] roots = abi.decode(_roots, (bytes32[2]));
```
follow previous examples that we have for parsing the proof to do it correctly.

File: contracts/anchors/bridged/AnchorPoseidon2.sol [hunk 5]
```diff
@@ -224,6 +223,25 @@ abstract contract AnchorPoseidon2 is MerkleTreePoseidon, ReentrancyGuard {
     }
   }
 
+  /** @dev */
+  function isKnownNeighborRoot(uint256 neighborChainID, bytes32 _root) public view returns (bool) {
```
Comment (drewstone) [inline contracts/anchors/bridged/AnchorPoseidon2.sol:227]: Please fix all spacing issues.

File: contracts/anchors/bridged/LinkableAnchorPoseidon2.sol [hunk 6]
```diff
@@ -40,7 +40,7 @@ abstract contract LinkableAnchorPoseidon2 is AnchorPoseidon2, ILinkableAnchor {
     bytes32[1] memory history = getLatestNeighborRoots();
     rootHistory[latestHistoryIndex] = history;
     // set the next history index modulo pruning length
-    latestHistoryIndex = latestHistoryIndex % pruningLength;
```
Comment (drewstone) [inline contracts/anchors/bridged/LinkableAnchorPoseidon2.sol:43]: afaik this function isn't being used anymore, right? We can remove it now that we have proper tracking of neighbor history and a new collection for the indices.

File: contracts/anchors/bridged/LinkableAnchorPoseidon2.sol [hunk 7]
```diff
@@ -80,6 +82,10 @@ abstract contract LinkableAnchorPoseidon2 is AnchorPoseidon2, ILinkableAnchor {
       root: root,
       height: height
     });
+     // add to root histories
+    currentNeighborRootIndex[sourceChainID] = (currentNeighborRootIndex[sourceChainID] + 1) % ROOT_HISTORY_SIZE;
+    uint32 neighborRootIndex = currentNeighborRootIndex[sourceChainID];
+    neighborRoots[sourceChainID][neighborRootIndex] = root;
```
Comment (drewstone) [inline contracts/anchors/bridged/LinkableAnchorPoseidon2.sol:88]: I would change this to first assign the neighborRootIndex.
```
    uint32 neighborRootIndex = (currentNeighborRootIndex[sourceChainID] + 1) % ROOT_HISTORY_SIZE;
    currentNeighborRootIndex[sourceChainID] = neighborRootIndex
    neighborRoots[sourceChainID][neighborRootIndex] = root;
```

**Review by drewstone — CHANGES_REQUESTED** [review]
Looks good but not exactly how we discussed it. The roots passed into the contract should be a string of bytes arbitrarily. We will parse it into the size we expect and so on and so forth.

File: test/helpers.js (hunk 1 above)
Comment (drewstone) [reply test/helpers.js:109]: what do you mean @nepoche ?

**Pull request merged**